### ISSUE TITLE
for loop error

### ISSUE THREAD

**Issue by sh0f** [vlang/v#5562]
**V version:** 0.1.28.1
**OS:** linux

**What did you do?**
```v
m := {'a': 'b', 'c': 'd'}
mut arr := []string{}
for k, _ in m {
	arr << k
}
println(arr)
```
**What did you expect to see?**
```v
['a', 'c']
```
**What did you see instead?**
```v
['
```

**Comment by yuyi98** [comment]
```v
m := {'a': 'b', 'c': 'd'}
mut arr := []string{}
for k, _ in m {
	arr << k.clone()
}
println(arr)
```
It can work.

**Comment by Ivo-Balbaert** [comment]
I'm seeing the same issue too; moreover on Windows, this code example outputs:
['P', 'c']     

I'm sure this is a regression. It worked on a V 0.1.28 of some days ago!

**Comment by spytheman** [comment]
I can confirm that is a regression. It works in 7be9526 and does:
```
0[10:34:06]delian@nemesis: /ssd2/opt/v_linux/v $ /dev/shm/v_at_7be9526/v run /v/vvvv/aa.v
['a', 'c']
```
I am still bisecting where it regressed.

**Comment by spytheman** [comment]
It regressed in 1848eb09 :
![regression commit](https://url4e.com/gyazo/images/6eb01c40.png)

**Comment by yuyi98** [comment]
There's a real problem in this, and this PR exposes the problem because the memory has been freed.

**Comment by spytheman** [comment]
I agree. The keys array is freed after the loop, but the values are used in the << operation.

**Comment by spytheman** [comment]
Yesterday, there was another bug that was fixed by doing `g.optionals << styp.clone()` too. I think the fix should be in the implementation of `<<`

**Comment by spytheman** [comment]
@yuyi98 why did you close https://github.com/vlang/v/pull/5592 ? I think you were on the right track to solving it. Was there another problem?

**Comment by yuyi98** [comment]
I add `array_clone` in it, but memory leak.

**Comment by spytheman** [comment]
with -autofree ?

**Comment by yuyi98** [comment]
no -autofree

**Comment by yuyi98** [comment]
I don't know how to solve this problem. Is it resolved in the parser stage?

**Comment by spytheman** [comment]
hm, you are right; doing an implicit clone will need much more work and analysis

**Comment by spytheman** [comment]
I think I have a working solution for the `for k,v in map {} case only.

**Comment by spytheman** [comment]
Fixed in ca1f47a :
![for k,v in map{} fix](https://url4e.com/gyazo/images/1cdfeb15.png)
